compiler/sem/sanitizer.nim

```diff
@@ -0,0 +1,976 @@
+        of nkElifBranch, nkElifExpr: parseBranch(c, it, true)
+        of nkElse, nkElseExpr:
+          guardLen(it, 1):
+            newTreeI(it.kind, it.info): expr(c, it[0])
+        else: invalidAst(c, it)
+  of nkTypeSection:
+    unreachable("missing")
+  of nkVarSection, nkLetSection:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] = parseVariable(c, n[i])
+  of nkConstSection:
+    unreachable("missing")
+  of nkConstDef, nkTypeDef:
+    # context dependent
+    invalid()
+  of nkDefer:
+    unreachable("missing")
+  of nkYieldStmt, nkRaiseStmt, nkReturnStmt, nkBreakStmt, nkContinueStmt, nkDiscardStmt:
+    checkSonsLen(n, 1)
+    result = prepareFrom(n)
+    result[0] =
+      if n[0].kind == nkEmpty: empty(n[0].info)
+      else:                    expr(c, n[0])
+
+  of nkFinally:
+    invalid()
+  of nkBlockStmt:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = stmt(c, n[1])
+  of nkStaticStmt:
+    checkSonsLen(n, 1)
+    result = stmt(c, n[0])
+  of nkStmtList:
+    result = stmts(c, n)
+  of nkImportStmt, nkImportExceptStmt, nkExportStmt, nkExportExceptStmt, nkFromStmt, nkIncludeStmt:
+    # QUESTION: should we perform any additional validation here or leave that
+    #           to sem?
+    checkMinSonsLen(n, 1)
+    result = exprs(c, n)
+  of nkBindStmt, nkMixinStmt:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] = qualifiedIdent(c, n[i])
+
+  of nkUsingStmt:
+    # we're much more strict than the parser here: everything besides an ident
+    # def is disallowed
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i in 0..<n.len:
+      result[i] =
+        case n.kind
+        of nkIdentDefs:   strictIdentDefs(c, n[i])
+        of nkCommentStmt: commentStmt(c, n[i])
+        else:             invalidAst(c, n[i])
+
+  of nkCommentStmt:
+    result = commentStmt(c, n)
+  of nkStmtListExpr:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in sliceIt(n, 0..^2):
+      result[i] = stmt(c, it)
+    result[^1] = expr(c, n[^1])
+  of nkBlockExpr:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = expr(c, n[1])
+  of nkStmtListType:
+    checkMinSonsLen(n, 1)
+    result = prepareFrom(n)
+    for i, it in sliceIt(n, 0..^2):
+      result[i] = stmt(c, it)
+    result[^1] = typeExpr(c, n[^1])
+  of nkBlockType:
+    checkSonsLen(n, 2)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+    result[1] = typeExpr(c, n[1])
+
+  of nkWith, nkWithout:
+    unreachable("missing")
+
+  of nkTypeOfExpr:
+    checkSonsLen(n, 1)
+    result = prepareFrom(n)
+    result[0] = expr(c, n[0])
+  of nkObjectTy,
+    nkTupleTy    ,
+    nkTupleClassTy,
+    nkTypeClassTy  ,
+    nkStaticTy:
+    unreachable("missing")
+  of nkRecList,
+    nkRecCase,
+    nkRecWhen:
+    # context dependent
+    invalid()
```

Comment:
If the `when` is eliminated then no worries, but if not, wouldn't we continue sanitizing recursively?

Reply:
The sanitizer logic for everything type definition related is not yet implemented, but once it is, both the condition and body AST of the `nkRecWhen` will also be considered.